README describing the workflow around external tests and their repositories

cameel opened the pull request:
Somewhat related to #10429 and all the chaos around updating branches in `solidity-external-tests`.

I decided to write down the recommended workflow since it's not all that obvious what can/should/must be done when updating these tests. Different people do it slightly differently and this way sometimes get in the way of each other. The current text describes how I think it should be done but it's also an opportunity to discuss it. I'll apply any changes that result from the discussion.

I'm not sure if a README in an obscure directory is the best place for it but I think it's not relevant enough to end-users to put it in the main docs. Someone tasked with updating these tests will probably manage to find it.

cameel committed:
README describing the workflow around external tests and their repositories

cameel commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+5. Discard the PRs and move the original branches to the place where the `_new` ones are and remove
+    the `_new` branches.
+
+#### Changes needed after a breaking release of the compiler
+When a breaking version of the compiler gets released and becomes the most recent release, the scripts
+and the branches in the forks need to be updated:
+- In each fork create a branch corresponding to the new brekaing version. E.g. if the current
+    branches are `master_070` and `master_080` the new one should be called `master_090`.
+- Leave the oldest (i.e. `master_070`) branch as is. We will not be updating it any more but there is no
+    need to remove it.
+- Update scripts on `develop` to now refer to the branch that used to be breaking (i.e. `master_080`)
+    and on `breaking` to refer to the newly added branch (i.e. `master_090`).
+    - Take care not to overwrite these changes when merging `develop` into breaking. Each branch
+        should always use the branches from the external repos that correspond to it.
```
Maybe this should be a part of the release checklist. Do we have one for breaking releases?

ekpyron replied:
I think in practice we'll only do it the first time an external test breaks, so since https://github.com/ethereum/solidity/pull/10556/files#r540011902 might be true, I don't think we need to put it in the release checklist.

cameel replied:
This is rewritten now. The new text suggests that it's fine to keep using the older branch and does not require doing anything on a breaking release..

cameel commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+    it not containing any of our customizations.
+2. If the update is straightforward and looks safe, go straight to point 5. Otherwise you need to
+    test it first.
+3. Create new branches corresponding to the two versioned ones but suffixed with `_new`. E.g.
+    `master_070_new` and `master_080_new`. Then rebase them so that they are on top of the updated
+    main branch. This may require tweaking some of the commits to apply our fixes in new places.
```
This rebasing does not play way well with submitting PRs to the fork repos (merge commits will be lost in the rebase and the code might disappear from PR page) and requires one doing it to have full access to modify the repo. But it makes it easier to keep the repos up to date with upstream.

hrkrshnn commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+
+Since projects often don't use the latest compiler, we keep a fork of each of these projects
+at https://github.com/solidity-external-tests/. If changes are needed to make a project work with the
+latest version of the compiler, they're maintained as a branch on top if the upstream master branch.
```
```suggestion
latest version of the compiler, they are maintained as a branch on top of the upstream master branch.
```

cameel replied:
Fixed.

hrkrshnn commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+3. Create new branches corresponding to the two versioned ones but suffixed with `_new`. E.g.
+    `master_070_new` and `master_080_new`. Then rebase them so that they are on top of the updated
+    main branch. This may require tweaking some of the commits to apply our fixes in new places.
+4. Create PRs on `deveop` and `breaking` in Solidity repo which modify the script to use the new
```
```suggestion
4. Create PRs on `develop` and `breaking` in Solidity repo which modify the script to use the new
```

cameel replied:
Fixed.

hrkrshnn commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+    the `_new` branches.
+
+#### Changes needed after a breaking release of the compiler
+When a breaking version of the compiler gets released and becomes the most recent release, the scripts
+and the branches in the forks need to be updated:
```
Do we really need to update it? If we merge breaking to develop, the scripts `colony.sh` etc should be the version from breaking.

cameel replied:
True. That's a great point! Given what @ekpyron also said about [not touching it unless something actually breaks](https://github.com/ethereum/solidity/pull/10556#discussion_r540181402), maybe it's better to have a system where an extra branch for a higher version is only created when it's actually neecessary and different from the earlier one. That will also save us some purely formal shuffling of branches in repos that still work without changes.

I'll change that in the description.

cameel replied:
Updated. Now it suggests that merging `breaking` into `develop` will resolve the issue automatically. I have also rewritten most of it to not require adding a separate branch for `breaking` unless necessary.

hrkrshnn reviewed (commented):
Looks good. 

Do we want to talk about the workflow when making changes for a breaking PR? Do we just push directly to the branches?

ekpyron commented:
> Do we want to talk about the workflow when making changes for a breaking PR? Do we just push directly to the branches?

(As you may have seen in the chainid PR) I usually create a branch and a draft PR in the external test repo, have a temporary commit in the solidity repo changing the tests to that branch, verify that it's ok, and then drop that commit from the solidity PR, marking all of them simultanously ready to merge.
I think that's reasonable. Note that depending on what we change, merging directly to the external test repos can easily break all other PRs until the corresponding solidity PR is merged, so better to try to do all of it at once in general.
If it's clear that the change is backwards-compatible, i.e. it won't break other test runs, and it's trivially correct, it's fine to skip all of that, though :-). Also not sure we need to document any of this.

ekpyron commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+This directory contains scripts for compiling some of the popular open-source projects using the
+current version of the compiler and running their test suites.
+
+Since projects often don't use the latest compiler, we keep a fork of each of these projects
```
Since @hrkrshnn started it ;-):
```suggestion
Since projects often do not use the latest compiler, we keep a fork of each of these projects
```
(personally I'd say it's fine, if not nicer, to contract, but IIRC it's policy not to in the repo)

cameel replied:
No worries, that's actually helpful. I shouldn't have left these in. I was going to make a final run through the file to find and eradicate these but it must have slipped my mind.

My problem is that these don't feel as casual to me as they apparently do to others so I just automatically use them everywhere and have to consciously keep myself from doing so. It's a downside of getting most of the exposure to English from what people write on the Internet I guess :)

ekpyron replied:
Actually from what I understand they actually aren't as casual as the policy in the repo may suggest - actually I myself rather know the opposite: I had native-English coauthors on scientific papers that *made me* use contractions :-). Still since the repo does it like that, we can keep it consistent.

cameel replied:
Interesting. Maybe it's regional?

Contraction removed.

ekpyron commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+latest version of the compiler, they're maintained as a branch on top if the upstream master branch.
+This is especially important for testing our `breaking` branch because we can not realistically expect
+external projects to be instantly compatible with a compiler version that has not been released yet.
+It also isolates us from upstream changes to some degree - their changes won't affect our test suite
+until we explicitly pull them.
```
Actually, I'd say that's a drawback in this, not a reason for this :-).
Also the main reason for doing it is to make sure our breaking changes are sane, by verifying that those projects *can* be upgraded.

cameel replied:
Is it really a drawback? There were cases like the new `util-contracts` in Gnosis v2 or Colony on 0.7.0 where we could get away with just using upstream directly but I thought that having random breakage in CI due to being exposed to their changes live would not be good so I created a fork or kept the branch. Should I change that back?

ekpyron replied:
I didn't mean to say we should change anything about how it works here (although I'd argue that we should always use the upstream version directly, if it works, precisely to have CI break and actually notice it breaking immediately rather than us missing issues because we insulated us. But that's up for debate and don't we do actually do that where possible anyways?)

What I rather meant here is that I wouldn't put forth isoluating us from upstream changes as reason for forking. Rather upstream not being up to date (wrt the supported solidity version) and us validating upgradability as sanity check for our breaking changes.

cameel replied:
I think it's good idea to have some isolation though. Upstream changes might not always be easy to adjust to and in the meantime our CI is left broken. Even more so if we start adding more of these projects. With, say, 100 projects there is a high chance of some project always being broken at any given time :)

I think that a good compromise might be to use the `master` branch of the fork (rather than e.g. `master_070`) whenever possible. In the workflow from the README, `master` is meant to be always kept in sync with the upstream without any special steps to protect against breakage on it, but if it breaks our tests it's easy to just move it back, fix the breakage and only then update again.

ekpyron commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+    `develop` and `breaking`. E.g. if the latest Solidity version is 0.7.5 and the main branch of the
+    external project is called `master`, create `master_070` and `master_080`. This is where we will
+    be adding our own commits. The one corresponding to the newer Solidity version should always be
+    on top on the older one. E.g. if a change is needed to keep the project
```
```suggestion
    on top of the older one. I.e. if a change is needed to keep the project
```
Maybe I'd also already say ``should always be *rebased* on top of the older one.``

By the way it's kind of sad that this situation ever occurs...

cameel replied:
I'll fix the `on` and `rebased` parts.

Regarding the `i.e.` vs `e.g.`, I actually meant the latter because it's an example using specific branch names.

> By the way it's kind of sad that this situation ever occurs...

I don't think we can expect anything else. Upgrades are extra work and they happen more often now. Some projects will always lag behind. It's good that at least some projects are moving away from 0.5.x now.

ekpyron commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+    If it is only needed for the compiler from its `breaking` branch, add it only in `master_080`.
+    - The fewer commits in these branches, the better. Ideally, any changes needed to make the compiler
+        work should be submitted upstream and the branches should just be tracking the upstream
+        one without any extra commits.
```
```suggestion
        ones without any extra commits.
```

ekpyron replied:
or ``the respective upstream one``. Although I'm starting to nitpick :-).

cameel replied:
I meant that there's one upstream branch (e.g. `master`) and multiple branches are tracking it (`master_070`, `master_080`).

We do have multiple upstream branches in some cases (`develop` and `v2` for Gnosis) but I didn't want to complicate the description by adding that. It should be pretty obvious how it works with multiple upstreams when you get how it does with one.

ekpyron replied:
Ok... I'd just drop the entire thing after the ``and`` then :-). But as you like :-)!

cameel replied:
I rewrote it like this:
```
Ideally, any changes needed to make the compiler work should be submitted upstream and our branches should just be tracking the main upstream branch without any extra commits.              
```

ekpyron commented on test/externalTests/README.md:
```diff
@@ -0,0 +1,66 @@
+#### Changes needed after a breaking release of the compiler
+When a breaking version of the compiler gets released and becomes the most recent release, the scripts
+and the branches in the forks need to be updated:
+- In each fork create a branch corresponding to the new brekaing version. E.g. if the current
```
```suggestion
- In each fork create a branch corresponding to the new breaking version. E.g. if the current
```

cameel replied:
Fixed.

ekpyron reviewed (commented):
In general, kudos, this is *really* specific and detailed, I think I wouldn't have had the patience :-). Definitely nice to have after the few minor nitpick-comments are straightend!

cameel committed:
fixup! README describing the workflow around external tests and their repositories

cameel committed:
fixup! README describing the workflow around external tests and their repositories

cameel commented:
@hrkrshnn 
> Do we want to talk about the workflow when making changes for a breaking PR? Do we just push directly to the branches?

I added a section describint that. It's basically what @ekpyron said above. I think that's a good workflow.

cameel commented:
Ready for another round. I have added a new section about PRs to breaking, rewrote the text not to require having a separate branch for `breaking` unless necessary (which means that we could have a 'hole' in versioning) and also mentioned that using on upstream branch that uses newer compiler is preferable when such a branch is available.

leonardoalt reviewed (approved):
Nice! Maybe @ekpyron or @hrkrshnn also want to take a look again
